### PR TITLE
Adapt the software UI to the new HTTP/JSON API

### PR DESCRIPTION
**Pull request by imobachgs**
The initial idea was to adapt the software page and the patterns selector to the new HTTP-based API. However, we took the chance to revamp the software page to make it more similar to other Agama parts.

<details>
  <summary>New software page</summary>

software-page
</details>

<details>
  <summary>New patterns selector</summary>

software-selection
</details>

## Changes summary

* Introduce a new software page that lists the patterns to install.
* Move the patterns selector to a pop-up.
* Introduce small changes to the software API.
* Update the JavaScript tests and enable the type checking on the new code.

## What is missing

- [ ] Distinguish between user and automatically selected patterns.

## Out of scope

* Improve the wording.

### CODE REVIEW

**Comment by mchf** [comment]
What is in rust & ruby LGTM

File: web/src/client/software.js [hunk 1]
```diff
@@ -52,6 +64,29 @@ const REGISTRATION_IFACE = "org.opensuse.Agama1.Registration";
...
 
+/**
+ * @typedef {object} SoftwareProposal
+ * @property {string} size - Used space in human-readable form.
+ * @property {Object.<string, number>} patterns - Selected patterns and the reason.
+ */
+
+/**
+ * @typedef {object} SoftwareConfig
+ * @property {Array<string>|undefined} patterns - An object where the keys are the pattern names
```
Comment (dgdavid) [inline web/src/client/software.js:75]: NP: something is wrong there. Type is an Array, but description mentions an object.

File: web/src/client/software.js [hunk 2]
```diff
@@ -52,6 +64,29 @@ const REGISTRATION_IFACE = "org.opensuse.Agama1.Registration";
...
+ */
+
+/**
+ * @typedef {Object} Pattern
+ * @property {string} name - pattern name (internal ID)
+ * @property {string} category - pattern category
+ * @property {string} summary - pattern name (user visible)
+ * @property {string} description - long description of the pattern
+ * @property {number} order - display order (string!)
+ * @property {string} icon - icon name (not path or file name!)
```
Comment (dgdavid) [inline web/src/client/software.js:87]: NP: different style from the rest, 

```suggestion
 * @property {string} name - Pattern name (internal ID).
 * @property {string} category - Pattern category.
 * @property {string} summary - User visible pattern name.
 * @property {string} description - Long description of the pattern.
 * @property {number} order - Display order (string!).
 * @property {string} icon - Icon name (not path or file name!).
```

File: web/src/client/software.js [hunk 3]
```diff
@@ -52,6 +64,29 @@ const REGISTRATION_IFACE = "org.opensuse.Agama1.Registration";
...
+ * @property {string|undefined} product - Product to install.
+ */
+
+/**
+ * @typedef {Object} Pattern
+ * @property {string} name - pattern name (internal ID)
+ * @property {string} category - pattern category
+ * @property {string} summary - pattern name (user visible)
+ * @property {string} description - long description of the pattern
+ * @property {number} order - display order (string!)
```
Comment (dgdavid) [inline web/src/client/software.js:86]: NP: type is number, but description says `string!`

File: web/src/client/software.js [hunk 4]
```diff
@@ -217,78 +196,75 @@ class SoftwareBaseClient {
...
    *
-   * @param {boolean} filter - `true` = filter the patterns, `false` = all patterns
-   * @return {Promise<Array<string>>}
+   * @return {Promise<Pattern[]>}
    */
-  async patterns(filter) {
-    const proxy = await this.client.proxy(SOFTWARE_IFACE);
-    return proxy.ListPatterns(filter);
+  async getPatterns() {
+    /** @type Array<{ name: string, category: string, summary: string, description: string, order: string, icon: string }> */
```
Comment (dgdavid) [inline web/src/client/software.js:230]: Question: Should we define this type elsewhere and the use it here?

File: web/src/client/software.js [hunk 5]
```diff
@@ -217,78 +196,75 @@ class SoftwareBaseClient {
...
-    return proxy.SelectedPatterns;
+  config() {
+    return this.client.get("/software/config");
   }
 
   /**
-   * Select a pattern to install
-   *
-   * @param {string} name - name of the pattern
+   * @param {Object.<string, boolean>} patterns - An object where the keys are the pattern names
```
Comment (dgdavid) [inline web/src/client/software.js:250]: NP: This looks like something I have read above where the type didn't match with the description. Same type or just a copy/paste mistake?

File: web/src/client/software.test.js [hunk 6]
```diff
@@ -33,44 +46,58 @@ const productProxy = {
...
   });
 
   describe("#getSelected", () => {
-    it("returns the selected product", async () => {
-      const client = new SoftwareClient();
-      const selectedProduct = await client.product.getSelected();
-      expect(selectedProduct).toEqual(
-        { id: "MicroOS", name: "openSUSE MicroOS" }
-      );
+    it.only("returns the selected product", async () => {
```
Comment (dgdavid) [inline web/src/client/software.test.js:95]: Question: this only is on purpose?

File: web/src/components/overview/SoftwareSection.jsx [hunk 7]
```diff
@@ -144,8 +144,7 @@ export default function SoftwareSection({ showErrors }) {
       icon="apps"
       loading={state.busy}
       errors={errors.map(toValidationError)}
-      // do not display the pattern selector when there are no patterns to display
-      path={Object.keys(state.patterns).length > 0 ? "/software" : null}
+      path="/software"
```
Comment (dgdavid) [inline web/src/components/overview/SoftwareSection.jsx:147]: :clap:

File: web/src/components/software/PatternSelector.jsx [hunk 8]
```diff
@@ -38,36 +35,10 @@ import { _ } from "~/i18n";
  * @property {string} description long description of the pattern
  * @property {string} order display order (string!)
  * @property {string} icon icon name (not path or file name!)
- * @property {number|undefined} selected who selected the pattern, undefined
+ * @property {number} selected who selected the pattern, undefined
```
Comment (dgdavid) [inline web/src/components/software/PatternSelector.jsx:38]: NP: using a different style here. I know it is following the previous style, just adding a note. Also, looks like it is not undefined anymore.

File: web/src/client/software.js (hunk 4 above)
Comment (imobachgs) [reply web/src/client/software.js:230]: Good question. I do not think it is needed, as it is the only place it is used. We convert those objects into our `Pattern` specification.

File: web/src/client/software.test.js (hunk 6 above)
Comment (imobachgs) [reply web/src/client/software.test.js:95]: Yes, it is on purpose. Most of the tests are related to the registration, and that's not implemented yet.

**Review by dgdavid — APPROVED** [review]
Let's go! Thanks 👍🏼

**Pull request merged**